File: javascript/test/v1-config-vectors.test.js
/**
 * Regression vectors for V1 transaction config (SIMD-0385) and
 * Reward.commissionBps (SIMD-0291).
 *
 * The base64 payloads are SubscribeUpdate messages encoded with
 * laserstream-core-proto 11.2.0 (Rust/prost) — the same bytes are asserted
 * against the Go and Rust SDKs, so all three decode identically.
 *
 * Run: npm run test:v1-config
 */

const assert = require('assert');
const { initProtobuf, decodeSubscribeUpdate } = require('../proto-decoder');

// SubscribeUpdate{transaction} whose message carries TransactionConfig
const V1_CONFIG_TX = 'CgR2ZWMxIpUCCowCCkAHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHGsMBCkAHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHEn8KBAgBGAISIAEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBEiACAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAhogCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkoAToPCJS05PTLAxDAuVUggIAQIgAoKhC72ZGsAQ==';
// SubscribeUpdate{block} whose rewards carry commissionBps
const COMMISSION_BPS_BLOCK = 'CgR2ZWMyKoYBCLzZkawBEgR0ZXN0GngKPgorVm90ZTExMTExMTExMTExMTExMTExMTExMTExMTExMTExMTExMTExMTExMRCIJxigjQYgBCoBNTIDNTUwCjYKK1N0YWtlMTExMTExMTExMTExMTExMTExMTExMTExMTExMTExMTExMTExMTEQj04YwJoMIAM=';

async function main() {
  await initProtobuf();

  // --- Vector 1: V1 transaction message with config ---
  const txUpdate = decodeSubscribeUpdate(Buffer.from(V1_CONFIG_TX, 'base64'));
  assert.ok(txUpdate.transaction, 'expected transaction update');
  assert.strictEqual(txUpdate.transaction.slot, '361000123');

  const msg = txUpdate.transaction.transaction.transaction.message;
  assert.strictEqual(msg.versioned, true);
  assert.ok(msg.config, 'expected message.config to be present for V1');
  assert.strictEqual(msg.config.priorityFee, '123456789012');
  assert.strictEqual(msg.config.computeUnitLimit, 1400000);
  assert.strictEqual(msg.config.heapSize, 262144);
  // Unset optional field must not be reported as present
  assert.ok(
    msg.config.loadedAccountsDataSizeLimit === undefined ||
      msg.config.loadedAccountsDataSizeLimit === null,
    'loadedAccountsDataSizeLimit must be absent'
  );

  // Legacy/V0-shaped messages must NOT grow a config
  const header = msg.header;
  assert.ok(header && header.numRequiredSignatures === 1);

  // --- Vector 2: block rewards with commissionBps ---
  const blockUpdate = decodeSubscribeUpdate(Buffer.from(COMMISSION_BPS_BLOCK, 'base64'));
  assert.ok(blockUpdate.block, 'expected block update');
  const rewards = blockUpdate.block.rewards.rewards;
  assert.strictEqual(rewards.length, 2);

  assert.strictEqual(rewards[0].pubkey, 'Vote111111111111111111111111111111111111111');
  assert.strictEqual(rewards[0].commission, '5');
  assert.strictEqual(rewards[0].commissionBps, '550');
  assert.strictEqual(rewards[0].rewardType, 4); // Voting

  // Second reward has no commission at all — empty string on the wire
  assert.strictEqual(rewards[1].commissionBps, '');
  assert.strictEqual(rewards[1].rewardType, 3); // Staking

  console.log('✅ v1-config vectors: all assertions passed');
}

main().catch((err) => {
  console.error('❌ v1-config vectors failed:', err.message);
  process.exit(1);
});

File: go/v1_config_vectors_test.go
package laserstream

// Regression vectors for V1 transaction config (SIMD-0385) and
// Reward.commission_bps (SIMD-0291).
//
// The base64 payloads are SubscribeUpdate messages encoded with
// laserstream-core-proto 11.2.0 (Rust/prost) — the same bytes are asserted
// against the JS and Rust SDKs, so all three decode identically.

import (
	"encoding/base64"
	"testing"

	pb "github.com/helius-labs/laserstream-sdk/go/proto"
	"google.golang.org/protobuf/proto"
)

// SubscribeUpdate{transaction} whose message carries TransactionConfig
const v1ConfigTxB64 = "CgR2ZWMxIpUCCowCCkAHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHGsMBCkAHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHEn8KBAgBGAISIAEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBEiACAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAhogCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkoAToPCJS05PTLAxDAuVUggIAQIgAoKhC72ZGsAQ=="

// SubscribeUpdate{block} whose rewards carry commission_bps
const commissionBpsBlockB64 = "CgR2ZWMyKoYBCLzZkawBEgR0ZXN0GngKPgorVm90ZTExMTExMTExMTExMTExMTExMTExMTExMTExMTExMTExMTExMTExMRCIJxigjQYgBCoBNTIDNTUwCjYKK1N0YWtlMTExMTExMTExMTExMTExMTExMTExMTExMTExMTExMTExMTExMTEQj04YwJoMIAM="

func mustDecode(t *testing.T, b64 string) *pb.SubscribeUpdate {
	t.Helper()
	raw, err := base64.StdEncoding.DecodeString(b64)
	if err != nil {
		t.Fatalf("bad base64: %v", err)
	}
	var update pb.SubscribeUpdate
	if err := proto.Unmarshal(raw, &update); err != nil {
		t.Fatalf("unmarshal: %v", err)
	}
	return &update
}

func TestV1TransactionConfigDecodes(t *testing.T) {
	update := mustDecode(t, v1ConfigTxB64)

	tx := update.GetTransaction()
	if tx == nil {
		t.Fatal("expected transaction update")
	}
	if tx.GetSlot() != 361000123 {
		t.Fatalf("slot = %d, want 361000123", tx.GetSlot())
	}

	msg := tx.GetTransaction().GetTransaction().GetMessage()
	if !msg.GetVersioned() {
		t.Fatal("expected versioned = true")
	}
	cfg := msg.GetConfig()
	if cfg == nil {
		t.Fatal("expected message.config to be present for V1")
	}
	if got := cfg.GetPriorityFee(); got != 123456789012 {
		t.Fatalf("priority_fee = %d, want 123456789012", got)
	}
	if got := cfg.GetComputeUnitLimit(); got != 1400000 {
		t.Fatalf("compute_unit_limit = %d, want 1400000", got)
	}
	if got := cfg.GetHeapSize(); got != 262144 {
		t.Fatalf("heap_size = %d, want 262144", got)
	}
	if cfg.LoadedAccountsDataSizeLimit != nil {
		t.Fatal("loaded_accounts_data_size_limit must be absent")
	}
}

func TestRewardCommissionBpsDecodes(t *testing.T) {
	update := mustDecode(t, commissionBpsBlockB64)

	block := update.GetBlock()
	if block == nil {
		t.Fatal("expected block update")
	}
	rewards := block.GetRewards().GetRewards()
	if len(rewards) != 2 {
		t.Fatalf("rewards = %d, want 2", len(rewards))
	}

	if got := rewards[0].GetCommissionBps(); got != "550" {
		t.Fatalf("commission_bps = %q, want \"550\"", got)
	}
	if got := rewards[0].GetCommission(); got != "5" {
		t.Fatalf("commission = %q, want \"5\"", got)
	}
	if rewards[0].GetRewardType() != pb.RewardType_Voting {
		t.Fatalf("reward_type = %v, want Voting", rewards[0].GetRewardType())
	}

	// Reward without commission: empty string on the wire
	if got := rewards[1].GetCommissionBps(); got != "" {
		t.Fatalf("commission_bps = %q, want empty", got)
	}
}
